pkg/tempopb/prealloc.go

```diff
@@ -0,0 +1,45 @@
+package tempopb
+
+import (
+	"github.com/prometheus/prometheus/pkg/pool"
+)
+
+var (
+	// buckets: [1KiB, 2KiB, 4KiB, 8KiB, 16KiB]
+	bytePool = pool.New(1_000, 16_000, 2, func(size int) interface{} { return make([]byte, 0, size) })
+)
+
+// PreallocRequest is a (repeated bytes requests) which preallocs slices on Unmarshal.
+type PreallocRequest struct {
+	Request []byte
+}
+
+// Unmarshal implements proto.Message.
+func (r *PreallocRequest) Unmarshal(dAtA []byte) error {
+	r.Request = bytePool.Get(len(dAtA)).([]byte)
+	r.Request = r.Request[:len(dAtA)]
+	copy(r.Request, dAtA)
+	return nil
+}
+
+// MarshalTo implements proto.Marshaller.
+func (r *PreallocRequest) MarshalTo(dAtA []byte) (int, error) {
+	copy(dAtA[:], r.Request[:])
+	return 0, nil // returned int is not used
```

Comment:
even if it's not used we should aim to fulfill the expectations of the interface